File: ports/pyagentic/agentic/conformance.py
"""Run the shared v1 conformance fixtures, read in place, against a `Runtime`.

    python -m agentic.conformance                 # every fixture, local runtime
    python -m agentic.conformance retry-tool      # one fixture by id
    python -m agentic.conformance --runtime NAME  # any registered runtime

Fixtures live in `spec/conformance/v1/fixtures/` of the Agentic-Streaming checkout. They
are located by walking up from this package or from `AGENTIC_SPEC_DIR`; not finding them
is an error, never a silent skip. A fixture whose `requires` includes a capability the
runtime does not declare `supported` or `partial` is recorded as skipped, never passed.
"""

from __future__ import annotations

import argparse
import os
import re
import sys
from concurrent.futures import Future
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Callable, Dict, List, Mapping, Optional, Protocol, Sequence, Union, runtime_checkable

import yaml

from .errors import AgenticError
from .events import Turn
from .runtime import LocalRuntime, Runtime, get_runtime


@runtime_checkable
class AsyncSubmitting(Protocol):
    """Runtimes that accept overlapping submissions (the fixtures' `concurrent_with`)."""

    def submit_async(self, event: Union[Turn, Mapping[str, Any]]) -> "Future[Dict[str, Any]]": ...


FIXTURE_SUBDIR = Path("spec") / "conformance" / "v1" / "fixtures"
RESULT_DETAIL_KEY = "runtime_detail"  # excluded from every comparison


class FixturesNotFound(AgenticError):
    error_class = "validation"


def fixtures_dir(start: Optional[Path] = None) -> Path:
    override = os.environ.get("AGENTIC_SPEC_DIR")
    if override:
        candidate = Path(override) / "conformance" / "v1" / "fixtures"
        if candidate.is_dir():
            return candidate
        raise FixturesNotFound(f"AGENTIC_SPEC_DIR={override!r} has no conformance/v1/fixtures directory")
    here = (start or Path(__file__)).resolve()
    for parent in [here] + list(here.parents):
        candidate = parent / FIXTURE_SUBDIR
        if candidate.is_dir():
            return candidate
    raise FixturesNotFound(
        "could not find spec/conformance/v1/fixtures above this package; run from an "
        "Agentic-Streaming checkout or set AGENTIC_SPEC_DIR=<checkout>/spec")


def fixture_paths(directory: Optional[Path] = None) -> List[Path]:
    paths = sorted((directory or fixtures_dir()).glob("*.yaml"))
    if not paths:
        raise FixturesNotFound(f"no fixtures in {directory or fixtures_dir()}")
    return paths


def load_yaml(path: Path) -> Dict[str, Any]:
    with open(path, encoding="utf-8") as fh:
        loaded: Dict[str, Any] = yaml.safe_load(fh)
    return loaded


def check_expectation(expected: Mapping[str, Any], actual: Mapping[str, Any]) -> List[str]:
    """The comparison rules of `spec/conformance/v1/README.md`. Returns mismatch messages."""
    problems: List[str] = []

    def mismatch(name: str, want: Any, got: Any) -> None:
        problems.append(f"{name}: expected {want!r}, got {got!r}")

    for name in ("conversation_id", "status", "path", "reply"):
        if name in expected and expected[name] != actual.get(name):
            mismatch(name, expected[name], actual.get(name))

    if "reply_matches" in expected:
        reply = actual.get("reply") or ""
        if not re.search(expected["reply_matches"], reply):
            mismatch("reply_matches", expected["reply_matches"], reply)

    if "error_class" in expected:
        got = (actual.get("error") or {}).get("class")
        if got != expected["error_class"]:
            mismatch("error_class", expected["error_class"], got)

    if "tool_calls" in expected:
        want = expected["tool_calls"]
        got = actual.get("tool_calls", [])
        if len(want) != len(got):
            mismatch("tool_calls length", len(want), len(got))
        else:
            for i, (w, g) in enumerate(zip(want, got)):
                if w["tool"] != g["tool"]:
                    mismatch(f"tool_calls[{i}].tool", w["tool"], g["tool"])
                for key in ("index", "attempt", "args"):
                    if key in w and w[key] != g.get(key):
                        mismatch(f"tool_calls[{i}].{key}", w[key], g.get(key))
                if w.get("failed", False) != (g.get("error") is not None):
                    mismatch(f"tool_calls[{i}].failed", w.get("failed", False), g.get("error"))

    types = [e["type"] for e in actual.get("events", [])]
    if "events_include" in expected:
        remaining = list(types)
        for wanted in expected["events_include"]:
            if wanted in remaining:
                remaining = remaining[remaining.index(wanted) + 1:]
            else:
                problems.append(f"events_include: {wanted} missing or out of order in {types}")
    for unwanted in expected.get("events_exclude", []):
        if unwanted in types:
            problems.append(f"events_exclude: {unwanted} present in {types}")

    for key, want in (expected.get("state_includes") or {}).items():
        got = (actual.get("state") or {}).get(key)
        if want != got:
            mismatch(f"state.{key}", want, got)

    return problems


@dataclass
class Outcome:
    fixture_id: str
    path: Path
    status: str  # pass | fail | skip
    problems: List[str] = field(default_factory=list)
    results: List[Dict[str, Any]] = field(default_factory=list)

    @property
    def reason(self) -> str:
        return "; ".join(self.problems)


RuntimeFactory = Callable[[], Runtime]


def turn_metadata(spec: Mapping[str, Any]) -> Dict[str, str]:
    """A fixture turn's `metadata` block as the string map a `Turn` carries."""
    return {str(k): str(v) for k, v in dict(spec.get("metadata") or {}).items()}


def run_fixture(path: Path, make_runtime: RuntimeFactory = LocalRuntime) -> Outcome:
    fixture = load_yaml(path)
    if fixture.get("workflow") is None:
        fixture["workflow"] = load_yaml((path.parent / fixture["workflow_ref"]).resolve())
    return run_fixture_document(fixture, make_runtime, path)


def run_fixture_document(fixture: Mapping[str, Any], make_runtime: RuntimeFactory = LocalRuntime,
                         path: Optional[Path] = None) -> Outcome:
    """Run one fixture whose `workflow` is already resolved and compare it with `expect`."""
    fixture_id = fixture["id"]
    path = path or Path(fixture_id)
    runtime = make_runtime()
    declared = runtime.capabilities()
    missing = sorted(cap for cap in fixture["requires"] if declared.get(cap) not in ("supported", "partial"))
    if missing:
        runtime.close()
        declared_as = ", ".join(f"{m}={declared.get(m, 'absent')}" for m in missing)
        return Outcome(fixture_id, path, "skip", [f"requires {missing}, declared {declared_as}"])

    workflow = fixture["workflow"]
    results: List[Dict[str, Any]] = []
    try:
        runtime.deploy(workflow)
        for batch in concurrent_batches(fixture["turns"]):
            if batch[0].get("restart_runtime"):
                runtime = _restart(runtime, workflow)
            results.extend(_deliver_batch(runtime, batch))
    except AgenticError as exc:
        return Outcome(fixture_id, path, "fail", [f"raised {type(exc).__name__}: {exc}"], results)
    finally:
        runtime.close()

    problems: List[str] = []
    for i, expected in enumerate(fixture["expect"]):
        if i >= len(results):
            problems.append(f"expect[{i}]: no result produced")
            continue
        problems += [f"expect[{i}] ({expected['turn_id']}) {p}" for p in check_expectation(expected, results[i])]
    return Outcome(fixture_id, path, "fail" if problems else "pass", problems, results)


def matrix_binding(fixture: Mapping[str, Any]) -> Union[List[Dict[str, Any]], Dict[str, str]]:
    """The `agentic.conformance` entry point for `spec/tools/conformance_matrix.py`.

    Receives a fixture with `workflow` resolved; returns the normalized results in turn order,
    or `{"skip": reason}` naming the capabilities the local runtime does not claim. The matrix
    runner does the comparison itself.
    """
    outcome = run_fixture_document(fixture)
    if outcome.status == "skip":
        return {"skip": outcome.reason}
    if len(outcome.results) < len(fixture["turns"]):
        raise AgenticError(outcome.reason)
    return outcome.results


def concurrent_batches(turns: Sequence[Mapping[str, Any]]) -> List[List[Mapping[str, Any]]]:
    """Consecutive turns joined by `concurrent_with` form one batch; every other turn is its own."""
    batches: List[List[Mapping[str, Any]]] = []
    i = 0
    while i < len(turns):
        turn = turns[i]
        group = set(turn.get("concurrent_with") or [])
        if not group:
            batches.append([turn])
            i += 1
            continue
        group.add(turn["turn_id"])
        j = i
        while j < len(turns) and turns[j]["turn_id"] in group:
            j += 1
        batches.append(list(turns[i:j]))
        i = j
    return batches


def _deliver_batch(runtime: Runtime, batch: Sequence[Mapping[str, Any]]) -> List[Dict[str, Any]]:
    """Deliver one batch of mutually `concurrent_with` turns: on a runtime with `submit_async`
    every turn is handed over in declared order without waiting for the previous one, so the
    turns are in flight together and the runtime decides how they overlap (only within their
    per-conversation order). The results come back in the declared order, whatever order the
    turns finished in. A runtime without `submit_async` receives the turns one after another."""
    turns = [Turn(conversation_id=spec["conversation_id"], turn_id=spec["turn_id"],
                  text=spec.get("text", ""), signal=spec.get("signal"),
                  metadata=turn_metadata(spec)) for spec in batch]
    if len(turns) == 1 or not isinstance(runtime, AsyncSubmitting):
        return [runtime.submit(turn) for turn in turns]
    futures = [runtime.submit_async(turn) for turn in turns]
    return [future.result() for future in futures]


def _restart(runtime: Runtime, workflow: Mapping[str, Any]) -> Runtime:
    restart = getattr(runtime, "restart", None)
    if not callable(restart):
        raise AgenticError(f"runtime {runtime.name!r} has no restart(); fixtures with restart_runtime "
                           f"cannot run against it")
    fresh = restart()
    if not isinstance(fresh, Runtime):
        raise AgenticError(f"{runtime.name}.restart() must return a Runtime")
    return fresh


def run_all(make_runtime: RuntimeFactory = LocalRuntime, only: Sequence[str] = (),
            directory: Optional[Path] = None) -> List[Outcome]:
    paths = fixture_paths(directory)
    if only:
        paths = [p for p in paths if load_yaml(p)["id"] in only]
        if not paths:
            raise FixturesNotFound(f"no fixture matches {list(only)}")
    return [run_fixture(p, make_runtime) for p in paths]


def main(argv: Optional[Sequence[str]] = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("ids", nargs="*", help="fixture ids to run (default: all)")
    parser.add_argument("--runtime", default="local")
    args = parser.parse_args(argv)
    outcomes = run_all(lambda: get_runtime(args.runtime), args.ids)
    counts = {"pass": 0, "fail": 0, "skip": 0}
    for outcome in outcomes:
        counts[outcome.status] += 1
        if outcome.status == "pass":
            print(f"pass {outcome.fixture_id}")
        elif outcome.status == "skip":
            print(f"skip {outcome.fixture_id}: {outcome.reason}")
        else:
            print(f"FAIL {outcome.fixture_id}")
            for problem in outcome.problems:
                print(f"     {problem}")
    print(f"\n{counts['pass']} passed, {counts['fail']} failed, {counts['skip']} skipped")
    return 1 if counts["fail"] else 0


if __name__ == "__main__":
    sys.exit(main())

File: agentic-clj/src/agentic/conformance.clj
(ns agentic.conformance
  "Binding of the shared v1 conformance fixtures (spec/conformance/v1/fixtures/*.yaml, read in place,
   never copied) to the Clojure runtime. A fixture is built with agentic.pipeline, its turns are
   delivered in order — honouring restart_runtime, signal and concurrent_with — and each normalized
   result is compared with the rules of spec/conformance/v1/README.md. A fixture whose `requires` we
   do not support is a skip, never a pass."
  (:require [clojure.java.io :as io]
            [clojure.string :as str]
            [agentic.spec :as spec]
            [agentic.pipeline :as pipeline]
            [agentic.core :as core]
            [agentic.log :as log]))

(def capabilities
  "What this runtime claims, in the terms of spec/v1/primitives.md."
  {"routing" :supported "rule_brain" :supported "tools" :supported
   "structured_tool_args" :supported "guardrails" :supported "verifier" :supported
   "ordering" :supported "idempotency" :supported "retry" :supported "memory" :supported
   "retrieval" :supported "replay" :supported "suspend_resume" :supported "saga" :supported
   "a2a" :supported "durable_store" :supported "context_window" :supported "llm_brain" :supported})

(defn fixtures-dir []
  (io/file (spec/spec-root) "conformance" "v1" "fixtures"))

(defn fixture-files []
  (->> (.listFiles (fixtures-dir))
       (filter #(re-find #"\.ya?ml$" (.getName ^java.io.File %)))
       (sort-by #(.getName ^java.io.File %))))

(defn read-fixture [^java.io.File f]
  (spec/read-document (.getPath f)))

(defn- workflow-of [fixture ^java.io.File f]
  (or (get fixture "workflow")
      (spec/read-document (.getPath (io/file (.getParentFile f) (get fixture "workflow_ref"))))))

(defn unsupported
  "The fixture's required capabilities this runtime does not claim as `supported`."
  [fixture]
  (vec (remove #(= :supported (get capabilities %)) (get fixture "requires"))))

;; ---- comparison rules ----

(defn- mismatch [field want got] (str field ": expected " (pr-str want) ", got " (pr-str got)))

(defn check-expectation
  "Failure messages for one expectation (fixture YAML map) against one wire-form result."
  [expected actual]
  (let [problems (atom [])
        fail! (fn [& args] (swap! problems conj (apply mismatch args)))]
    (doseq [field ["conversation_id" "status" "path" "reply"]]
      (when (and (contains? expected field) (not= (get expected field) (get actual field)))
        (fail! field (get expected field) (get actual field))))
    (when-let [re (get expected "reply_matches")]
      (let [reply (or (get actual "reply") "")]
        (when-not (re-find (re-pattern re) reply) (fail! "reply_matches" re reply))))
    (when (contains? expected "error_class")
      (let [got (get-in actual ["error" "class"])]
        (when (not= got (get expected "error_class")) (fail! "error_class" (get expected "error_class") got))))
    (when-let [want (get expected "tool_calls")]
      (let [got (get actual "tool_calls" [])]
        (if (not= (count want) (count got))
          (fail! "tool_calls length" (count want) (count got))
          (doseq [[i w g] (map vector (range) want got)]
            (when (not= (get w "tool") (get g "tool")) (fail! (str "tool_calls[" i "].tool") (get w "tool") (get g "tool")))
            (doseq [k ["index" "attempt" "args"]]
              (when (and (contains? w k) (not= (get w k) (get g k))) (fail! (str "tool_calls[" i "]." k) (get w k) (get g k))))
            (when (not= (boolean (get w "failed" false)) (some? (get g "error")))
              (fail! (str "tool_calls[" i "].failed") (get w "failed" false) (get g "error")))))))
    (let [types (mapv #(get % "type") (get actual "events" []))]
      (loop [remaining types [wanted & more] (get expected "events_include")]
        (when wanted
          (let [i (.indexOf ^java.util.List remaining wanted)]
            (if (neg? i)
              (do (swap! problems conj (str "events_include: " wanted " missing or out of order in " (pr-str types)))
                  (recur remaining more))
              (recur (subvec remaining (inc i)) more)))))
      (doseq [unwanted (get expected "events_exclude")]
        (when (some #{unwanted} types)
          (swap! problems conj (str "events_exclude: " unwanted " present in " (pr-str types))))))
    (doseq [[k want] (get expected "state_includes")]
      (let [got (get-in actual ["state" k])]
        (when (not= want got) (fail! (str "state." k) want got))))
    @problems))

;; ---- delivery ----

(defn- ->event [turn]
  {:conversation-id (get turn "conversation_id")
   :turn-id (get turn "turn_id")
   :user-id (get turn "user_id" "anonymous")
   :text (get turn "text" "")
   :signal (get turn "signal")})

(defn- deliver-batch
  "Deliver a group of mutually `concurrent_with` turns: all are enqueued (arrive) in declared order
   without waiting for any to finish, so they are in flight together and the runtime must serialize
   them per conversation; the results come back in the declared order."
  [system turns]
  (let [outcomes (mapv #(core/submit-async system (->event %)) turns)]
    (mapv (fn [p] (let [r @p] (if (instance? Throwable r) (throw r) r))) outcomes)))

(defn- batches
  "Consecutive turns joined by `concurrent_with` form one batch; every other turn is its own."
  [turns]
  (loop [[t & more] turns out []]
    (cond
      (nil? t) out
      (seq (get t "concurrent_with"))
      (let [group (into #{(get t "turn_id")} (get t "concurrent_with"))
            [batch rest] (split-with #(group (get % "turn_id")) (cons t more))]
        (recur rest (conj out (vec batch))))
      :else (recur more (conj out [t])))))

(defn run-fixture
  "Run one fixture file. Returns {:id :status (:passed|:failed|:skipped) :problems [...] :results [...]}."
  [^java.io.File f]
  (let [fixture (read-fixture f)
        id (get fixture "id")
        missing (unsupported fixture)]
    (if (seq missing)
      {:id id :status :skipped :problems [(str "requires " (pr-str missing))]}
      (let [workflow (spec/load-workflow (workflow-of fixture f))
            stores (pipeline/open-stores workflow)
            system (atom (pipeline/system-from workflow stores))
            results (reduce (fn [acc batch]
                              (when (get (first batch) "restart_runtime")
                                (swap! system #(pipeline/system-from workflow (select-keys % [:store :state :log]))))
                              (into acc (deliver-batch @system batch)))
                            [] (batches (get fixture "turns")))
            wire (mapv log/->wire results)
            problems (vec (mapcat (fn [i expected]
                                    (let [actual (get wire i)
                                          errs (if actual
                                                 (into (spec/result-problems actual) (check-expectation expected actual))
                                                 [(str "no result for expectation " i)])]
                                      (map #(str "expect[" i "] " %) errs)))
                                  (range) (get fixture "expect")))]
        {:id id :status (if (seq problems) :failed :passed) :problems problems :results wire}))))

(defn run-all
  "Every fixture in spec order."
  []
  (mapv run-fixture (fixture-files)))

(defn report
  "A one-line-per-fixture summary."
  [outcomes]
  (str/join "\n" (map (fn [{:keys [id status problems]}]
                        (str (name status) " " id (when (seq problems) (str " — " (str/join "; " problems)))))
                      outcomes)))
